File: solutions/cpp/luhn/1/luhn.cpp
#include "luhn.h"

namespace luhn {
    bool valid(const std::string num)
    {
        int sum = 0;
        int digits = 0;
        int parity = 1;
        for (int i = num.size() - 1; i >= 0; i--)
        {
            if (num[i] == ' ' || num[i] == '\t')
                continue ;
            if (num[i] < '0' || num[i] > '9')
                return false;
            digits++;
            int digit = (num[i] - '0') * parity;
            if (digit > 9)
                digit -= 9;
            parity = 3 - parity;
            sum += digit;
        }
        return digits > 1 && sum % 10 == 0;
    }
}  // namespace luhn

File: solutions/javascript/luhn/1/luhn.js
//
// This is only a SKELETON file for the 'Luhn' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export const valid = digitString => {
  let noSpaces = digitString.replace(/\s/g, "")
  if (!(/^\d+$/.test(noSpaces))) {
    return false;
  }
  let sum = 0;
  for (let i = noSpaces.length - 1; i >= 0; i--) {
    if ((i + noSpaces.length) % 2 == 0) {
      sum += 2 * +noSpaces[i];
      if (+noSpaces[i] > 4) {
        sum -= 9;
      }
    } else {
      sum += +noSpaces[i];
    }
  }
  return noSpaces.length > 1 && sum % 10 == 0;
};
